### PR TITLE
Visual integration with LSP 

### PR DESCRIPTION
**Pull request by akalburgi**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 1]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
...
     where optionalPrefix = "$/"
 
+filesFromExecParams :: List Aeson.Value -> [NormalizedFilePath]
+filesFromExecParams (List files) = map (toNormalizedFilePath . T.unpack) (concatMap fileStringToPath files)
+            where fileStringToPath :: Aeson.Value -> [T.Text]
+                  fileStringToPath (Aeson.String x) = [x]
+                  fileStringToPath (Aeson.Array arr) = concatMap fileStringToPath arr
+                  fileStringToPath _ex = error ("Failed to get daml files from workspace" ++ show _ex )
+
+onCommand
```
Comment (akalburgi) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: May we should move it to a different module?

File: compiler/damlc/daml-visual/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Copyright (c) 2019 The DAML Authors. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+load("//bazel_tools:haskell.bzl", "da_haskell_library")
```
Comment (akalburgi) [inline compiler/damlc/daml-visual/BUILD.bazel]: has to move out of lib as it was complaining about cyclic dependency

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 3]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
...
+        Just path -> do
+            case filesFromExecParams path of
+                (head : rest) -> do
+                        logInfo (ideLogger ide) "Generating visualization for current daml project"
+                        mbmodules <- mapM (\f -> runAction ide (useWithStale GenerateDalf f)) rest
+                        Just (WhnfPackage package, _) <- runAction ide (useWithStale GeneratePackage head)
+                        pkgMap <- runAction ide  (useNoFile_ GeneratePackageMap)
+                        let extpkgs = map dalfPackagePkg $ Map.elems pkgMap
+                        let wrld = LF.initWorldSelf extpkgs package
+                        let modules = map (fst . fromJust ) mbmodules
```
Comment (rohanjr) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: As discussed, might be worth having a comment here as to why the `Nothing` case should not happen.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 4]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
...
+                        let dots = T.pack $ Visual.dotFileGen modules wrld
+                        return $ Aeson.String dots
+                _ -> return $ Aeson.String "Could not consutruct world and the module list."
```
Comment (rohanjr) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]:
```suggestion
                _ -> return $ Aeson.String "Could not construct world and the module list."
```

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 5]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
...
+    -> IO Aeson.Value
+onCommand ide ExecuteCommandParams{..} = do
+    case _arguments of
+        Nothing -> return $ Aeson.String "No .daml files where found in the IDE workspace"
+        Just path -> do
+            case filesFromExecParams path of
+                (head : rest) -> do
+                        logInfo (ideLogger ide) "Generating visualization for current daml project"
+                        mbmodules <- mapM (\f -> runAction ide (useWithStale GenerateDalf f)) rest
+                        Just (WhnfPackage package, _) <- runAction ide (useWithStale GeneratePackage head)
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: The pattern match is not safe, see below for an explanation of when this can be `Nothing`.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 6]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
...
+                        pkgMap <- runAction ide  (useNoFile_ GeneratePackageMap)
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: If you change the visualization to just accept a world you should be able to simplify this a fair bit an donly have one call to `GeneratePackage` and one to `GeneratePackageMap`.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 7]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
     -- $/setTraceNotification which we want to ignore.
     where optionalPrefix = "$/"
 
+filesFromExecParams :: List Aeson.Value -> [NormalizedFilePath]
+filesFromExecParams (List files) = map (toNormalizedFilePath . T.unpack) (concatMap fileStringToPath files)
+            where fileStringToPath :: Aeson.Value -> [T.Text]
```
Comment (hurryabit) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: The name `fileStringToPath` is very misleading. Maybe something like `collectTexts` is better.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 8]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
...
+onCommand
+    :: IdeState
+    -> ExecuteCommandParams
+    -> IO Aeson.Value
+onCommand ide ExecuteCommandParams{..} = do
+    case _arguments of
+        Nothing -> return $ Aeson.String "No .daml files where found in the IDE workspace"
+        Just path -> do
+            case filesFromExecParams path of
+                (head : rest) -> do
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: If you go for option 3 as I described above, you can also avoid the weird special treatment of the first file which seems rather confusing.

File: compiler/daml-extension/src/extension.ts [hunk 9]
```diff
@@ -99,34 +99,21 @@ function getViewColumnForShowResource(): ViewColumn {
...
-            }
-            else {
-                exec(visualizeCmd, execOpts, ((error: Error, stdout: string, stderr: string) => {
-                    if (error) {
-                        vscode.window.showErrorMessage("damlc visual command failed with " + error)
-                    }
-                    else {
-                        vscode.workspace.openTextDocument({ content: stdout, language: "dot" })
+function visualize() {
+    if (vscode.window.activeTextEditor) {
```
[the remark this answers is not in the document]
Comment (akalburgi) [reply compiler/daml-extension/src/extension.ts]: Now displaying and INFO if no DAML file is open

File: compiler/daml-extension/src/extension.ts [hunk 10]
```diff
@@ -99,34 +99,21 @@ function getViewColumnForShowResource(): ViewColumn {
...
+        const workspace = vscode.workspace.workspaceFolders![0];
+        if (workspace) {
+            vscode.workspace.findFiles(new vscode.RelativePattern(workspace, "**/*.daml"), "").then(r => {
```
[the remark this answers is not in the document]
Comment (akalburgi) [reply compiler/daml-extension/src/extension.ts]: Went with 3rd approach we now generate visual for open files only.

File: compiler/daml-extension/src/extension.ts [hunk 11]
```diff
@@ -99,34 +99,21 @@ function getViewColumnForShowResource(): ViewColumn {
...
+                let files = r.map(f => f.path)
+                damlLanguageClient.sendRequest(ExecuteCommandRequest.type,
+                    { command: "daml/damlVisualize", arguments: [files] }).then(r => {
```
[the remark this answers is not in the document]
Comment (akalburgi) [reply compiler/daml-extension/src/extension.ts]: :( renamed that to dotFileContents

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 12]
```diff
@@ -68,6 +77,39 @@ setIgnoreOptionalHandlers = PartialHandlers $ \WithMessage{..} x -> return x
...
+                        logInfo (ideLogger ide) "Generating visualization for current daml project"
+                        mbmodules <- mapM (\f -> runAction ide (useWithStale GenerateDalf f)) rest
```
[the remark this answers is not in the document]
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: We do not need this anymore but I did know there was a thing to run in parallel thanks

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs (hunk 6 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: Yup now it is reduced to using one`GeneratePackage` and one `GeneratePackageMap`

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs (hunk 8 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: Yup this is simplified now

File: compiler/daml-extension/src/extension.ts [hunk 13]
```diff
@@ -99,34 +99,21 @@ function getViewColumnForShowResource(): ViewColumn {
...
-                exec(visualizeCmd, execOpts, ((error: Error, stdout: string, stderr: string) => {
-                    if (error) {
-                        vscode.window.showErrorMessage("damlc visual command failed with " + error)
-                    }
-                    else {
-                        vscode.workspace.openTextDocument({ content: stdout, language: "dot" })
+function visualize() {
+    if (vscode.window.activeTextEditor) {
+        const workspace = vscode.workspace.workspaceFolders![0];
+        if (workspace) {
```
[the remark this answers is not in the document]
Comment (akalburgi) [reply compiler/daml-extension/src/extension.ts]: Now displaying info if we do not have a file open

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs (hunk 5 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: No more doing this

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs (hunk 3 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: Fixed

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs (hunk 1 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: moved it to `DA.Daml.LanguageServer.Visualize`

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs (hunk 7 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: Simplified further we will not need that function

File: compiler/daml-extension/src/extension.ts [hunk 14]
```diff
@@ -99,34 +99,19 @@ function getViewColumnForShowResource(): ViewColumn {
...
-                            .then(doc => vscode.window.showTextDocument(doc, vscode.ViewColumn.One, true)
-                                .then(_ => loadPreviewIfAvailable()))
-                    }
-                }))
-            }
-        }))
-
-    }))
+function visualize() {
+    if (vscode.window.activeTextEditor){
```
Comment (cocreature) [inline compiler/daml-extension/src/extension.ts]: What happens if I have an open window but it’s not a .daml file?

File: compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs [hunk 15]
```diff
@@ -248,6 +249,18 @@ generatePackageRule =
         dalf <- use_ GenerateDalf file
         return ([], Just $ WhnfPackage $ deps{LF.packageModules = NM.insert dalf (LF.packageModules deps)})
 
+generateVisualizationRule :: Rules ()
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs]: I don’t think adding a rule for visualization is a good idea:

Adding more rules is the usual memory vs time tradeoff. If you put it in a rule, you get caching so you don’t have to recompute it if nothing has changed. On the other hand you have to keep it in memory which increases memory usage.

In this case caching is quite poor: As soon as any file in the dependency tree of `file` has changed you will rerun the whole visualization. That means that the only case where we benefit from caching is if the user reruns the visualization without making any changes. That does not seem like a very common case that we should be optimizing for. Compare that to a rule like typechecking that reuses the typechecked modules so you gain a lot more from caching since while the user might have modified the current file at least some of the dependencies will not have changed so caching is crucial to give fast feedback on incremental changes.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 16]
```diff
@@ -27,12 +27,14 @@ import Development.IDE.Core.Rules
...
 import Development.IDE.Core.Service.Daml
 
+import DA.Daml.LanguageServer.Visualize
 import qualified Network.URI                               as URI
 
 import Language.Haskell.LSP.Messages
 import qualified Language.Haskell.LSP.Core as LSP
 
 
+
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: Please don’t include unnecessary whitespace changes. They only clutter diffs.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs [hunk 17]
```diff
@@ -101,11 +105,11 @@ runLanguageServer
...
-    let handlers = setHandlersKeepAlive <> setHandlersVirtualResource <> setHandlersCodeLens <> setIgnoreOptionalHandlers
+    let handlers = setHandlersKeepAlive <> setHandlersVirtualResource <> setHandlersCodeLens <> setIgnoreOptionalHandlers <> setCommandHandler
     LS.runLanguageServer options handlers getIdeState
 
 
 options :: LSP.Options
 options = def
     { LSP.codeLensProvider = Just $ CodeLensOptions $ Just False
-    }
+    }
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: Why did you remove the trailing newline?

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs [hunk 18]
```diff
@@ -0,0 +1,56 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+-- | Execute Daml commands
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: That description tells you less than the module name which at least mentions `Visualize`. Either make it more useful or just remove it.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+module DA.Daml.LanguageServer.Visualize
+    ( setCommandHandler
+    ) where
+
+import qualified Data.Aeson as Aeson
+import           Language.Haskell.LSP.Types
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: Please don’t indent align newly added imports.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+import           Development.IDE.LSP.Server
+
+import Development.IDE.Types.Logger
+
+import qualified Data.Text as T
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: I don’t understand the reasoning behind grouping and sorting the imports this way. Unless there is a good reason to split up imports (e.g. to differentiate between imports from external libraries and our own code), I prefer to just have a single block of imports sorted alphabetically by the module name.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    ExecuteCommandParams "daml/damlVisualize" (Just _arguments) -> do
+        case collectTexts _arguments of
+            Just mod -> do
+                    logInfo (ideLogger ide) "Generating visualization for current daml project"
+                    Just dots <- runAction ide (use GenerateVisualization mod)
+                    return $ Aeson.String dots
+            Nothing     -> do
+                logError (ideLogger ide) "Expected a single module to visualize, got multiple module"
+                return $ Aeson.String "Expected a single module to visualize, got multiple module"
+    ExecuteCommandParams  _ (Just _arguments) -> do
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: Command handlers are expected to be extensible. If you get commands that you cannot handle, just ignore them.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+onCommand ide execParsms = case execParsms of
+    ExecuteCommandParams "daml/damlVisualize" (Just _arguments) -> do
+        case collectTexts _arguments of
+            Just mod -> do
+                    logInfo (ideLogger ide) "Generating visualization for current daml project"
+                    Just dots <- runAction ide (use GenerateVisualization mod)
+                    return $ Aeson.String dots
+            Nothing     -> do
+                logError (ideLogger ide) "Expected a single module to visualize, got multiple module"
+                return $ Aeson.String "Expected a single module to visualize, got multiple module"
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: This is still not going via the error mechanism in LSP. It looks like doing that will require modifications to `withResponse` so I’m fine with doing that in a separate PR but please add a `TODO`.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+        logError (ideLogger ide) "Command is not supported"
+        return Aeson.Null
+    ExecuteCommandParams  _ Nothing -> do
+        logError (ideLogger ide) "Missing DAML module to visualize"
+        return Aeson.Null
+
+setCommandHandler ::PartialHandlers
+setCommandHandler = PartialHandlers $ \WithMessage{..} x -> return x {
+    LSP.executeCommandHandler = withResponse RspExecuteCommand $ const onCommand
+}
```
Comment (cocreature) [inline compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: Missing trailing newline.

File: compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs (hunk 15 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs]: Ah ok that makes sense, I did not think it about memory consumption.

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs (hunk 21 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: `PartialHandlers` is expected to be extensible but I think we will have to handle all the ExecuteCommands once we have `ExecuteCommandHandler`. I am not sure how we can ignore without getting incomplete match warning

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs (hunk 20 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: I was following the same style within the modules in that folder. I have no strong preference here. Will get rid of it

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs (hunk 22 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: sounds good. Done

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs (hunk 17 above)
Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer.hs]: :( Not sure adding it back

File: compiler/daml-extension/src/extension.ts (hunk 14 above)
Comment (akalburgi) [reply compiler/daml-extension/src/extension.ts]: Now should display an info message to open a DAML file

File: compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs (hunk 21 above)
Comment (cocreature) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: Good point, we cannot handle commands that partially handle a request at the moment. That will need some refactoring. I would suggest that you combine the two cases here (the first where you have arguments but the wrong command and the second where you don’t have arguments into a single case and log both the command name and the arguments.

Comment (akalburgi) [reply compiler/damlc/daml-ide/src/DA/Daml/LanguageServer/Visualize.hs]: Cool. Will do that

File: compiler/lsp-tests/src/Main.hs [hunk 24]
```diff
@@ -416,6 +417,44 @@ scenarioTests run = testGroup "scenarios"
...
+            , "  where"
+            , "    signatory owner"
+            , "    controller owner can"
+            , "      Delete : ()"
+            , "        do return ()"
+            ]
+        Just escapedFp <- pure $ uriToFilePath (main' ^. uri)
+        actualDotString :: ExecuteCommandResponse <- LSP.request WorkspaceExecuteCommand $ ExecuteCommandParams
+           "daml/damlVisualize"  (Just (List [Aeson.String $ T.pack escapedFp]))
+        let expectedDotString = "digraph G {\ncompound=true;\nrankdir=LR;\nsubgraph cluster_Coin{\nn0[label=Create][color=green]; \nn1[label=Archive][color=red]; \nn2[label=Delete][color=red]; \nlabel=Coin;color=blue\n}\n}\n"
```
Comment (akalburgi) [inline compiler/lsp-tests/src/Main.hs]: This looks more like golden test. Let me know if there is a better way to do this.

File: compiler/lsp-tests/src/Main.hs [hunk 25]
```diff
@@ -416,6 +417,44 @@ scenarioTests run = testGroup "scenarios"
...
+    , testCase "Invalid commands result in empty response"  $ run $ do
+        main' <- openDoc' "Main.daml" damlId $ T.unlines
+            [ "daml 1.2"
+            , "module Empty where"
+            ]
+        Just escapedFp <- pure $ uriToFilePath (main' ^. uri)
+        actualDotString :: ExecuteCommandResponse <- LSP.request WorkspaceExecuteCommand $ ExecuteCommandParams
+           "daml/NoCommand"  (Just (List [Aeson.String $ T.pack escapedFp]))
+        let expectedNull = Just Aeson.Null
+        liftIO $ assertBool "Invlalid command" (expectedNull == _result actualDotString)
```
Comment (cocreature) [inline compiler/lsp-tests/src/Main.hs]:
```suggestion
        liftIO $ assertEqual "Invalid command" (Just Aeson.Null) (_result actualDotString)
```

File: compiler/lsp-tests/src/Main.hs [hunk 26]
```diff
@@ -416,6 +417,44 @@ scenarioTests run = testGroup "scenarios"
...
+        closeDoc main'
+    , testCase "Visualization command with no arguments" $ run $ do
+        actualDotString :: ExecuteCommandResponse <- LSP.request WorkspaceExecuteCommand $ ExecuteCommandParams
+           "daml/damlVisualize"  Nothing
+        let expectedNull = Just Aeson.Null
+        liftIO $ assertBool "Invlalid command" (expectedNull == _result actualDotString)
```
Comment (cocreature) [inline compiler/lsp-tests/src/Main.hs]:
```suggestion
        liftIO $ assertEqual "Invlalid command" (Just Aeson.Null) (_result actualDotString)
```
